Specifying one associated type makes Rust forget about the constraint on the other ones

I tried this code:

```rust
use std::marker::PhantomData;

pub struct XImpl<T, E, F2, F1>
where
	F1: FnMut() -> T,
	F2: Fn(E),
{
	f1: F1,
	f2: F2,
	_ghost: PhantomData<E>,
}

pub trait X<T>: Sized {
	type F1: FnMut() -> T;
	type F2: Fn(Self::E);
	type E;

	fn and<NewF1, NewF1Generator>(self, f: NewF1Generator) -> XImpl<T, Self::E, Self::F2, NewF1>
	where
		NewF1: FnMut() -> T,
		NewF1Generator: FnOnce(Self::F1) -> NewF1;
}

impl<T, E, F2, F1> X<T> for XImpl<T, E, F2, F1>
where
	F1: FnMut() -> T,
	F2: Fn(E),
{
	type E = E;
	type F2 = F2;
	type F1 = F1;

	fn and<NewF1, NewF1Generator>(self, f: NewF1Generator) -> XImpl<T, E, F2, NewF1>
	where
		NewF1: FnMut() -> T,
		NewF1Generator: FnOnce(F1) -> NewF1,
	{
		XImpl {
			f1: f(self.f1),
			f2: self.f2,
			_ghost: PhantomData,
		}
	}
}

fn f() -> impl X<()> {
	XImpl {
		f1: || (),
		f2: |()| (),
		_ghost: PhantomData,
	}
}

fn f2() -> impl X<()> {
	f().and(|mut rb| move || rb())
}
```

It does compile: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2a998f955ecf30d5debe97b01f2658df

If however we specify the values for the associated type `E` in `f` and `f2`, so that code calling `f2()` can depend on knowing the type `E`:
```rust
fn f() -> impl X<(), E = ()> {
	XImpl {
		f1: || (),
		f2: |()| (),
		_ghost: PhantomData,
	}
}

fn f2() -> impl X<(), E = ()> {
	f().and(|mut rb| move || rb())
}
```
The compiler seems to forget that any `X::F2` implements `Fn(Self::E)`, and stops compiling: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=11918c893cff872958b75b3c4a84dfb0
```
   Compiling playground v0.0.1 (/playground)
error[E0277]: expected a `std::ops::Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
  --> src/lib.rs:55:6
   |
55 |     f().and(|mut rb| move || rb())
   |         ^^^ expected an `Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
   |
   = help: the trait `std::ops::Fn<((),)>` is not implemented for `<impl X<()> as X<()>>::F2`

error[E0277]: expected a `std::ops::Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
  --> src/lib.rs:55:27
   |
13 | pub trait X<T>: Sized {
   | --------------------- required by `X`
...
55 |     f().and(|mut rb| move || rb())
   |                              ^^^^ expected an `Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
   |
   = help: the trait `std::ops::Fn<((),)>` is not implemented for `<impl X<()> as X<()>>::F2`

error[E0277]: expected a `std::ops::Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
  --> src/lib.rs:54:12
   |
54 | fn f2() -> impl X<(), E = ()> {
   |            ^^^^^^^^^^^^^^^^^^ expected an `Fn<((),)>` closure, found `<impl X<()> as X<()>>::F2`
55 |     f().and(|mut rb| move || rb())
   |     ------------------------------ this returned value is of type `XImpl<(), (), <impl X<()> as X<()>>::F2, [closure@src/lib.rs:55:19: 55:31 rb:_]>`
   |
   = help: the trait `std::ops::Fn<((),)>` is not implemented for `<impl X<()> as X<()>>::F2`
   = note: required because of the requirements on the impl of `X<()>` for `XImpl<(), (), <impl X<()> as X<()>>::F2, [closure@src/lib.rs:55:19: 55:31 rb:_]>`
   = note: the return type of a function must have a statically known size
```

### Meta
`rustc --version --verbose`:
```
rustc 1.43.1 (8d69840ab 2020-05-04)
binary: rustc
commit-hash: 8d69840ab92ea7f4d323420088dd8c9775f180cd
commit-date: 2020-05-04
host: x86_64-unknown-linux-gnu
release: 1.43.1
LLVM version: 9.0
```

This also happens on nightly.

--- a/tests/ui/impl-trait/associated-type-fn-bound-issue-72207.rs
+++ b/tests/ui/impl-trait/associated-type-fn-bound-issue-72207.rs
@@ -1,0 +1,57 @@
+//@ check-pass
+//@ compile-flags: --crate-type=lib
+
+#![allow(dead_code)]
+
+use std::marker::PhantomData;
+
+pub struct XImpl<T, E, F2, F1>
+where
+    F2: Fn(E),
+{
+    f1: F1,
+    f2: F2,
+    _ghost: PhantomData<(T, E)>,
+}
+
+pub trait X<T>: Sized {
+    type F1;
+    type F2: Fn(Self::E);
+    type E;
+
+    fn and<NewF1, NewF1Generator>(self, f: NewF1Generator) -> XImpl<T, Self::E, Self::F2, NewF1>
+    where
+        NewF1Generator: FnOnce(Self::F1) -> NewF1;
+}
+
+impl<T, E, F2, F1> X<T> for XImpl<T, E, F2, F1>
+where
+    F2: Fn(E),
+{
+    type E = E;
+    type F2 = F2;
+    type F1 = F1;
+
+    fn and<NewF1, NewF1Generator>(self, f: NewF1Generator) -> XImpl<T, E, F2, NewF1>
+    where
+        NewF1Generator: FnOnce(F1) -> NewF1,
+    {
+        XImpl {
+            f1: f(self.f1),
+            f2: self.f2,
+            _ghost: PhantomData,
+        }
+    }
+}
+
+fn f() -> impl X<(), E = ()> {
+    XImpl {
+        f1: || (),
+        f2: |()| (),
+        _ghost: PhantomData,
+    }
+}
+
+fn f2() -> impl X<(), E = ()> {
+    f().and(|rb| rb)
+}

